### PR TITLE
fix(pythonrt): close grpc connection

### PR DESCRIPTION
**Pull request by ig-ra**
ensure grpc connection is closed in error/success scenarios

### CODE REVIEW

**Comment by itayd** [comment]
would appreciate an explanation of the fix in the description

File: runtimes/pythonrt/runner_manager.go [hunk 1]
```diff
@@ -12,13 +12,23 @@ import (
...
-type runnerType string
+type (
+	runnerType   string
+	RunnerClient struct {
+		pb.RunnerClient
+		cc *grpc.ClientConn
+	}
+)
+
+func (c *RunnerClient) Close() {
```
Comment (tebeka) [inline runtimes/pythonrt/runner_manager.go]: Return the error from `c.cc.Close()`

File: runtimes/pythonrt/pythonrt.go [hunk 2]
```diff
@@ -76,6 +76,7 @@ func (py *pySvc) cleanup(ctx context.Context) {
 	if err := runnerManager.Stop(ctx, py.runnerID); err != nil {
 		py.log.Warn("close runner", zap.Error(err))
 	}
+	py.runner.Close()
```
Comment (tebeka) [inline runtimes/pythonrt/pythonrt.go]: See below about returning error, log it here.

File: runtimes/pythonrt/runner_manager.go [hunk 3]
```diff
@@ -54,16 +64,17 @@ func waitForServer(name string, h Healther, timeout time.Duration) error {
...
-func dialRunner(addr string) (pb.RunnerClient, error) {
+func dialRunner(addr string) (*RunnerClient, error) {
 	creds := insecure.NewCredentials()
 	conn, err := grpc.NewClient(addr, grpc.WithTransportCredentials(creds))
 	if err != nil {
 		return nil, err
 	}
 
-	c := pb.NewRunnerClient(conn)
+	c := &RunnerClient{pb.NewRunnerClient(conn), conn}
```
Comment (tebeka) [inline runtimes/pythonrt/runner_manager.go]: I prefer to use pointer only when returning, this way it clear on return that we use pointer semantics.

```
c := RunnerClient{pb.NewRunnerClient(conn), conn}
...

return &c, nil
```

File: runtimes/pythonrt/runner_manager.go [hunk 4]
```diff
@@ -54,16 +64,17 @@ func waitForServer(name string, h Healther, timeout time.Duration) error {
...
 
 	if err := waitForServer("runner", c, 10*time.Second); err != nil {
+		conn.Close()
```
Comment (tebeka) [inline runtimes/pythonrt/runner_manager.go]: Check error

**Review by tebeka — CHANGES_REQUESTED** [review]

File: runtimes/pythonrt/runner_manager.go (hunk 1 above)
Comment (ig-ra) [reply runtimes/pythonrt/runner_manager.go]: done

File: runtimes/pythonrt/pythonrt.go (hunk 2 above)
Comment (ig-ra) [reply runtimes/pythonrt/pythonrt.go]: done

File: runtimes/pythonrt/runner_manager.go (hunk 4 above)
Comment (ig-ra) [reply runtimes/pythonrt/runner_manager.go]: we are in the middle of handling other (more important error).
Do you want me to join errors?

File: runtimes/pythonrt/runner_manager.go (hunk 3 above)
Comment (ig-ra) [reply runtimes/pythonrt/runner_manager.go]: done

File: runtimes/pythonrt/runner_manager.go (hunk 4 above)
Comment (tebeka) [reply runtimes/pythonrt/runner_manager.go]: Sure, join them.

**Review by tebeka — APPROVED** [review]

File: runtimes/pythonrt/runner_manager.go (hunk 4 above)
Comment (ig-ra) [reply runtimes/pythonrt/runner_manager.go]: done

**Pull request merged**